### PR TITLE
Support import sourt source with postgresql

### PR DESCRIPTION
**Pull request by raviks789**
Earlier import source was not working because of the grouping function. Here this has been fixed.

### CODE REVIEW

File: library/X509/ProvidedHook/HostsImportSource.php [hunk 1]
```diff
@@ -9,22 +9,39 @@ class HostsImportSource extends x509ImportSource
...
+                    'host_ip'       => 't.ip',
+                    'host_name'     => 't.hostname',
+                    'host_ports'    => 'GROUP_CONCAT(DISTINCT t.port SEPARATOR ",")'
+                ])
+                ->join('x509_certificate_chain cc', 'cc.id = t.latest_certificate_chain_id')
+                ->join('x509_certificate_chain_link ccl', 'ccl.certificate_chain_id = cc.id')
+                ->join('x509_certificate c', 'c.id = ccl.certificate_id')
+                ->where(['ccl.order = ?' => 0])
+                ->groupBy(['t.ip', 't.hostname']);
+        }
```
Comment (nilmerg) [inline library/X509/ProvidedHook/HostsImportSource.php:38]: Please duplicate less code here. You can for example do the following:

```php
$targets = (new ...)->from(...);

if ($postgres) {
    $targets->columns(...);
} else {
    $targets->columns(...);
}

$targets->join(...)
    ->join(...)
    ->...;

```

File: library/X509/ProvidedHook/HostsImportSource.php [hunk 2]
```diff
@@ -9,22 +9,39 @@ class HostsImportSource extends x509ImportSource
...
 
-        $results = [];
         $foundDupes = [];
         foreach ($this->getDb()->select($targets) as $target) {
+            if ($this->getDb()->getConfig()->host === 'postgres') {
+                $target->host_ip = hex2bin($target->host_ip);
+            }
```
Comment (nilmerg) [inline library/X509/ProvidedHook/HostsImportSource.php:44]: Please add an empty line after closing braces.

File: library/X509/ProvidedHook/HostsImportSource.php [hunk 3]
```diff
@@ -9,22 +9,39 @@ class HostsImportSource extends x509ImportSource
...
+                    'host_ports'    => 'GROUP_CONCAT(DISTINCT t.port SEPARATOR ",")'
+                ])
+                ->join('x509_certificate_chain cc', 'cc.id = t.latest_certificate_chain_id')
+                ->join('x509_certificate_chain_link ccl', 'ccl.certificate_chain_id = cc.id')
+                ->join('x509_certificate c', 'c.id = ccl.certificate_id')
+                ->where(['ccl.order = ?' => 0])
+                ->groupBy(['t.ip', 't.hostname']);
+        }
 
-        $results = [];
```
Comment (nilmerg) [inline library/X509/ProvidedHook/HostsImportSource.php:25]: Please re-add this. It's still in use after all.

**Review by nilmerg — CHANGES_REQUESTED** [review]
There's also the `ServicesImportSource` which may also be subject to this issue. Please change this as well.

File: library/X509/ProvidedHook/ServicesImportSource.php [hunk 4]
```diff
@@ -9,36 +9,69 @@ class ServicesImportSource extends x509ImportSource
...
+                    'cert_subject'          => 'c.subject',
+                    'cert_issuer'           => 'c.issuer',
+                    'cert_self_signed'      => 'COALESCE(ci.self_signed, c.self_signed)',
+                    'cert_trusted'          => 'c.trusted',
+                    'cert_valid_from'       => 'c.valid_from',
+                    'cert_valid_to'         => 'c.valid_to',
+                    'cert_fingerprint'      => 'HEX(c.fingerprint)',
+                    'cert_dn'               => 'GROUP_CONCAT(CONCAT(dn.key, \'=\', dn.value) SEPARATOR \',\')',
+                    'cert_subject_alt_name' => (new Sql\Select())
+                        ->from(['x509_certificate_subject_alt_name can'])
```
Comment (nilmerg) [inline library/X509/ProvidedHook/ServicesImportSource.php:56]: The array here is redundant. That the implementation of `from()` does it isn't a reason why we should do it here. Please undo this change.

File: library/X509/ProvidedHook/ServicesImportSource.php [hunk 5]
```diff
@@ -9,36 +9,69 @@ class ServicesImportSource extends x509ImportSource
...
+                ->joinLeft('x509_dn dn', 'dn.hash = c.subject_hash')
+                ->where(['ccl.order = ?' => 0])
+                ->groupBy(['t.ip', 't.hostname', 't.port']);
+        }
 
         $results = [];
         foreach ($this->getDb()->select($targets) as $target) {
+            if ($this->getDb()->getConfig()->db === 'pgsql') {
+                $target->host_ip = hex2bin($target->host_ip);
+            }
```
Comment (nilmerg) [inline library/X509/ProvidedHook/ServicesImportSource.php:74]: Please also add an empty line after this closing brace.

**Review by nilmerg — CHANGES_REQUESTED** [review]
Tested it now myself successfully. Results from MySQL and PostgreSQL are identical for me, so it works fine as it should! :+1: 

But please apply the changes I've requested, now two additional ones.

**Comment by nilmerg** [comment]
And please rebase with master. This then ensures the CI checks will run.

**Comment by lippserd** [comment]
Should be pushed to #85.

**Comment by raviks789** [comment]
> Should be pushed to #85.

The changes have been pushed to #85

**Pull request closed without merging**